### ISSUE TITLE
PBC get_bands() uses a lot of memory for large systems

### ISSUE THREAD

**Issue by lkwagner** [pyscf/pyscf#719]
get_bands() is written as follows:
```
       fock = self.get_hcore(cell, kpts_band)
        fock = fock + self.get_veff(cell, dm_kpts, kpts=kpts, kpts_band=kpts_band)
        s1e = self.get_ovlp(cell, kpts_band)
        mo_energy, mo_coeff = self.eig(fock, s1e)
        if single_kpt_band:
            mo_energy = mo_energy[0]
            mo_coeff = mo_coeff[0]
        return mo_energy, mo_coeff
```

This appears to allocate many objects that scale with the number of k-points. But since this is not self-consistent, it seems like you could instead break up the k-points when the total memory gets big. This can be done by the user, but it'd be nice not to have to!

**Comment by tberkel** [comment]
For each k-point, `mo_coeff` will be the same size as `fock` and `s1e`.  So
if the user doesn't want to put the latter two in memory for a long list of
k-points, then presumably they don't want to have the long list of
`mo_coeff` returned to them in memory?  (Sure, there's a factor of 3
difference ...)

**Comment by lkwagner** [comment]
Yes, I think that's right. It's very common just to want the energies, not the coefficients for every k-point.

**Comment by sunqm** [comment]
The memory usage in this function is about 3*Nkpts*Nao^2. Excluding the orbital coefficients at each k point, at least ~2*Nkpts*Nao^2 for the fock and s1e matrices. It is not a big save in memory. Actually the size of Nkpts*Nao^2 is not terribly large. For example, even 10000 k points with Nao=500, the memory requirement is about ~40 GB memory. If you get memory issue, it is likely caused by the get_veff function for a long list of k points. @lkwagner do you still think return orbital coefficients in get_bands consumes too much memory?

**Comment by lkwagner** [comment]
I don't think that 40 GB is a small amount of memory, especially when the memory requirement *could* be 10,000 times smaller. And if NAO is bigger then it rises quite quickly. 500 basis functions is not that many. We certainly did have examples back then where we had to break it up into sections. 

Just from a user expectation, I would not expect `get_bands()` to return orbital coefficients. I would expect only energies, since `bands` means the energies.

I would expect coefficients from something like `get_ncsf_orbitals()` or something like that.
